dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,662 @@
+	// Standard object metadata
+	// +optional
+	metav1.ObjectMeta `json:"metadata,omitempty" protobuf:"bytes,1,opt,name=metadata"`
+
+	ResourceClaimSpecAlternatives `json:",inline"` // Inlined with "spec" defined in the ResourceClaimSpec.
+
+	// Status describes whether the claim is ready for use.
+	// +optional
+	Status ResourceClaimStatus `json:"status,omitempty" protobuf:"bytes,3,opt,name=status"`
 }
 
-// DeviceClaimSpec details the requirements that devices chosen
-// to satisfy this claim must meet.
-type DeviceClaimSpec struct {
-	// MatchAttributes allows specifying a constraint that will apply
-	// across all of the claims. For example, if you specified "numa", then
-	// this overall claim could only be successfully fulfilled if all of
-	// the included claims could be fulfilled by pools with the same "numa"
-	// attribute value. If we simply set matchAttributes in each claim
-	// separately, then they could be consistent within claims, but
-	// inconsistent across claims. Therefore, we need this additional
-	// constraint.
+// ResourceClaimSpecAlternatives defines how a claim is to be allocated.
+type ResourceClaimSpecAlternatives struct {
+	// Spec define what to allocated and how to configure it.
+	// Spec and SpecRef are mutually exclusive.
+	// +optional
+	Spec *ResourceClaimSpec `json:"spec"`
+
+	// SpecRef references a separate object with the specification of the claim
+	// that will be used by the driver when allocating a resource for the
+	// claim. Parameters and ParametersRef are mutually exclusive.
+	//
+	// SpecRef is typically used to reference a vendor CR.
+	// A vendor controller then converts that CR into a ResourceClaimSpecification
+	// object and that is then used to allocate the claim.
 	//
+	// The object must be in the same namespace as the ResourceClaim.
 	// +optional
+	SpecRef *ResourceClaimSpecReference `json:"specRef,omitempty"`
+}
+
+// Used inside a ResourceClaimSpecAlternatives or a ResourceClaimSpecification object.
+type ResourceClaimSpec struct {
+	// ResourceClassName references additional configuration and requirements
+	// that apply to the whole claim and all requests in it. If the class
+	// contains default requests, then those are used if (and only if)
+	// the claim does not provide those itself.
+	//
+	// Requirements in the class must be satisfied in addition to the
+	// requirements in the claim.
+	//
+	// +optional
+	ResourceClassName string `json:"resourceClassName,omitempty" protobuf:"bytes,1,name=resourceClassName"`
+
+	// To be decided: does it make sense to allow referencing multiple
+	// classes?  As it stands now, there could be a need to create a large
+	// variety of different classes where each class is one combination of
+	// different options.
+	//
+	// When allowing different classes, each class could describe one aspect:
+	// - "devices from vendor foo"
+	// - "more than 10 GiB of RAM"
+	//
+	// Probably need better use cases for classes. Stuff for a new KEP in >= 1.32...
```

Comment:
Except that adding a `resourceClasses` field would be a breaking change because it would get ignored by older schedulers. Perhaps we should already allow it now and replace `resourceClassName string` with `resourceClasses []string`?

Reply:
The use case for allowing multiple classes is, for example "I limit devices to SRIO-V pf NICs (first class) from vendor foozer (second class)".

Reply:
I went for a list of one-ofs and no separate field for a single name:
```
requests:
- classes:
  - name: example.com-foozer
  - name: same-pci-root
```

See `dra-evolution/testdata/pod-one-container-shared-split-allocation-gpus.yaml`.

The `same-pci-root` class adds a vendor-independent match attribute:
```
# Whatever devices get allocated, they must be on the same PCI root.
apiVersion: resource.k8s.io/v1alpha
kind: ResourceClass
metadata:
  name: same-pci-root
claim:
  requirements:
  - match:
      attribute: pcie-root.dra.k8s.io
```

For this to work, I removed the restriction that referencing a class in a request doesn't inherit claim requirements and config. I think inheriting both makes sense and might be expected, as in this example.

See 7703dfa for the change.

Reply:
Making the `classes` entries a one-of enables adding other kinds of classes in the future.

Reply:
To keep the simple case simple, we could add `className: example.com-foozer`.

I'm undecided whether `className` and `classes` then should be usable together (implied ordering: first `className`, then `classes`) or mutually exclusive.

Reply:
I'm still not sold on the idea of creating a bunch of small, "atomic" classes and then composing them together inside the claim like this. I understand why one might want to do it, but it feels very error prone and hard to debug if you get it wrong. At least with a single class I only have one object I need to look at in order to reconcile with any "local" constraints / requirements my claim has. With what's being proposed here, you would need to reconcile across `n` objects, keeping their order of inheritance in mind.

Reply:
To be clear -- I'm not saying I can't be convinced otherwise -- I just foresee it causing more problems than it solves.

I also feel that this basically reduces `ResourceClasses` to a set of "helm-like template functions", and (as such) why should admins be the only ones that have the ability to create these?

It also feels like a "bastardization" of what a `ResourceClass` was initially intended for, which is maybe where my general uneasiness with it comes from. Meaning that it feels heavy-handed to have a bunch of top-level k8s objects written solely as syntactic-sugar for hiding the complexity of a few (possibly-disjoint-but-not-guaranteed-to-be) CEL expressions.

Reply:
> I'm still not sold on the idea of creating a bunch of small, "atomic" classes and then composing them together inside the claim like this.

I'm not sure either. But I'm also not sure that a single class will be sufficient and that exactly the `ResourceClass` that we are designing right now is the ultimate answer.

Given these uncertainties, I'm leaning towards providing flexibility (allow multi-inheritance) and future extension points (the one-of [`ClassReference`](https://github.com/kubernetes-sigs/wg-device-management/blob/c920b2be9077bdb0d0ba6369e1d6d0f89239982c/dra-evolution/pkg/api/claim_types.go#L293-L303)).

If you all feel that I am overly cautious here, then let's drop multi-inheritance. But can we at least keep the `ClassReference`?

That means in YAML:
```yaml
class:
   name: example.com-foozer
```

> I also feel that this basically reduces ResourceClasses to a set of "helm-like template functions", and (as such) why should admins be the only ones that have the ability to create these?

Because if users were writing them, they could put the same content also directly into their claim. If it's repetitive, then tools like kustomize or other templating tools can help. We don't need in-tree support.

Reply:
By dropping the multi-inheritance and inlining the `ClassReference` we are back at a simple `resourceClassName` field. The only difference compared to 1.30 is that it's optional and provides a future extension point.

I think the consensus is that multi-inheritance doesn't carry its own weight, so I'll remove that. Then we can continue discussing whether even that single reference is unnecessary.

Reply:
See https://github.com/kubernetes-sigs/wg-device-management/pull/14/commits/1dbb6d270abc7a32489f959d73ab6982ff7d94c5